### PR TITLE
[#63] [Android] [Integrate] As a user, I can submit my answers

### LINKED ISSUE #63: [Android] [Integrate] As a user, I can submit my answers
# Why

A user must be able to submit their answers when finished with all the questions.

# Acceptance Criteria

- Call the endpoint which is implemented in #13 to submit answers.
- Show the `Thank you` screen (#59)

### PR DESCRIPTION
**Pull request by AVI5HEK**
- Close #63 

## What happened 👀

- Call the endpoint which is implemented in https://github.com/nimblehq/avishek-phong-kmm-ic/issues/13 to submit answers.
- Show the Thank you screen (https://github.com/nimblehq/avishek-phong-kmm-ic/issues/59)

## Insight 📝

N/A.

## Proof Of Work 📹

https://github.com/nimblehq/avishek-phong-kmm-ic/assets/8093908/71564040-593c-4135-a201-59cb6a3ecd0d

### Request body
```json
    {
        "survey_id": "d5de6a8f8f5f1cfe51bc",
        "questions": [
            {
                "id": "940d229e4cd87cd1e202",
                "answers": [
                    {
                        "id": "4cbc3e5a1c87d99bc7ee",
                        "answer": "1"
                    }
                ]
            },
            {
                "id": "ea0555f328b3b0124127",
                "answers": [
                    {
                        "id": "b31dca3a711cdddc06da",
                        "answer": "2"
                    }
                ]
            },
            {
                "id": "16e68f5610ef0e0fa4db",
                "answers": [
                    {
                        "id": "534f04bca9d051c8deb3",
                        "answer": "3"
                    }
                ]
            },
            {
                "id": "bab38ad82eaf22afcdfe",
                "answers": [
                    {
                        "id": "ae5825c1b2acf846d53a",
                        "answer": "4"
                    }
                ]
            },
            {
                "id": "85275a0bf28a6f3b1e63",
                "answers": [
                    {
                        "id": "46c607dbf315a7d1790f",
                        "answer": "5"
                    }
                ]
            },
            {
                "id": "642770376f7cd0c87d3c",
                "answers": [
                    {
                        "id": "c5e8fc3bca388ec15884",
                        "answer": "1"
                    }
                ]
            },
            {
                "id": "b093a6ad9a6a466fa787",
                "answers": [
                    {
                        "id": "002d338bc37f2142ad44",
                        "answer": "TripAdvisor"
                    },
                    {
                        "id": "4ac9af2d4bc4a4dd6315",
                        "answer": "Website"
                    },
                    {
                        "id": "20907898b2c2d68159c8",
                        "answer": "Staying at hotel"
                    },
                    {
                        "id": "2051cda19e6940173c9f",
                        "answer": "Other"
                    }
                ]
            },
            {
                "id": "e593b2fa2f81891a2b1e",
                "answers": [
                    {
                        "id": "e90a40f6fcf2b9a8e593",
                        "answer": "5"
                    }
                ]
            },
            {
                "id": "c3a9b8ce5c2356010703",
                "answers": [
                    {
                        "id": "2a49e148c5b170aca804",
                        "answer": "I'm not sure if you have any questions or concerns please visit the plug-in settings in the first time in the first time in the "
                    }
                ]
            },
            {
                "id": "fbf5d260de1ee6195473",
                "answers": [
                    {
                        "id": "491d49dd6b8174456acf",
                        "answer": "John"
                    },
                    {
                        "id": "6db6dcae1a6c6644d723",
                        "answer": "000000000000"
                    },
                    {
                        "id": "575db8c074601994bde3",
                        "answer": "john@doe.com"
                    }
                ]
            },
            {
                "id": "4372463ce56db58c0983",
                "answers": [
                ]
            }
        ]
    }
```

### CODE REVIEW

**Review by suho — APPROVED** [review]

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt [hunk 1]
```diff
@@ -3,44 +3,35 @@ package co.nimblehq.avishek.phong.kmmic.presentation.module
...
 class SurveyQuestionViewModel(
     private val submitSurveyAnswerUseCase: SubmitSurveyAnswerUseCase,
-    private val dispatchersProvider: DispatchersProvider
+    private val dispatchersProvider: DispatchersProvider,
 ) : BaseViewModel() {
 
     private val _viewState: MutableStateFlow<SurveyQuestionViewState> =
         MutableStateFlow(SurveyQuestionViewState())
-    val viewSate: StateFlow<SurveyQuestionViewState> = _viewState
+    val viewState: StateFlow<SurveyQuestionViewState> = _viewState
```
Comment (phongvhd93) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt:26]: Would you help me to fix the issue related to my typo? The fix is replacing `SurveyQuestionCombineViewModel:25` in iOS module with this `createPublisher(for: sharedViewModel.viewStateNative)` 🙏

Comment (AVI5HEK) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt:26]: @phongvhd93 done https://github.com/nimblehq/avishek-phong-kmm-ic/pull/139/commits/34aa0f91ac2cecbe7b0c17ba05c2703414abd7af

**Review by phongvhd93 — APPROVED** [review]

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt [hunk 2]
```diff
@@ -50,8 +51,13 @@ fun SurveyDetailScreen(
     }
     var shouldShowExitConfirmationDialog by remember { mutableStateOf(false) }
 
+    LaunchedEffect(surveyDetailViewState.survey) {
+        surveyDetailViewState.survey?.let(surveyQuestionViewModel::updateStateWith)
+    }
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:56]: @AVI5HEK Can you remind me why are we observing state like this?

Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:56]: @kaungkhantsoe  due to the way iOS implemented the passing of the `survey` from one screen to another. Basically, we're showing the intro and the questions on the same screen, but iOS is doing it in separate screens. So they're passing the object whereas we're observing it. Perhaps @phongvhd93  can shed some light on this.

Comment (phongvhd93) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:56]: When a user taps on the `Start Survey` we will fetch the survey detail and pass it to the view model of the next screen. If there is an error, the user will stay at the `Survey Detail` screen. We won't fetch questions in the question screen, so passing the survey is required.

Comment (kaungkhantsoe) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:56]: @phongvhd93 The thing is if I comment that out it still works for questions but it won't submit the surveys.

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt [hunk 3]
```diff
@@ -30,12 +29,14 @@ const val ImageScaleAnimationDurationInMillis = 700
 fun SurveyDetailScreen(
     homeViewModel: HomeViewModel = getViewModel(),
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:30]: Why are we getting `HomeviewModel` in `SurveyDetailScreen`? We can get the survey detail information from the API, can't we? 💭

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt (hunk 2 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:56]: @kaungkhantsoe that's correct. We need the other VM for submitting the survey answers.

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt (hunk 3 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:30]: @kaungkhantsoe Fetching the survey has some latency. But clicking the survey on the Home screen magnifies the image and displays the survey title and description immediately on the Details screen. So the screens share the same VM to use the already fetched data.

**Pull request merged**